niceman/distributions/debian.py

```diff
@@ -457,31 +462,22 @@ def _find_all_sources(self):
                     archive_uri=src_vals.get("archive_uri"))
 
     def _get_pkgs_arch_and_version(self, pkg_dicts):
-        # Define a function to process batches of packages and return
-        # an array of parsed dpkg -s output
-        def proc_dpkg_cmd(subqueries, cmd_results):
-            try:
-                out, _ = self._session.execute_command(
-                    ['dpkg', '-s'] + subqueries
-                )
-                out = utils.to_unicode(out, "utf-8")
-                # dpkg -s uses the same output as apt-cache show pkg
-                cmd_results += parse_apt_cache_show_pkgs_output(out)
-            except CommandError as _:
-                raise  # Currently raise any exception we encounter
-            return cmd_results
-
         # Convert package names to name:arch format
         queries = []
         for p in pkg_dicts:
             # Use "dpkg -s pkg" to get the installed version and arch
             # Note: "architecture" is in the dict, but may be null
             queries.append(p["name"] if not p["architecture"]
                            else "%(name)s:%(architecture)s" % p)
-        # Determine how big our batches can be
-        num_files = get_cmd_batch_len(queries, 8)
-        # Now process our file list and return the file to package mapping
-        results = batch_process_list(proc_dpkg_cmd, queries, num_files, [])
+        # Call "dpkg -s" in batches
+        exec_gen = execute_command_batch(self._session, ['dpkg', '-s'],
+                                         queries)
+        # Parse and accumulate "dpkg -s" results in an array
+        results = []
+        for (out, _, _) in exec_gen:
+            out = utils.to_unicode(out, "utf-8")
+            # dpkg -s uses the same output as apt-cache show pkg
+            results += parse_apt_cache_show_pkgs_output(out)
```

Comment:
all good, but if you want to become a really pythonic fella you could express yourself via a list comprehension.
```python
results = [
   parse_apt_cache_show_pkgs_output(utils.to_unicode(out, "utf-8"))
   for (out, _, _) in exec_gen
]
```
and then a very pythonic fella would have made it into a generator ( `()` instead of `[]`) to not even incur temporary list construction if  `self.create_lookup_from_apt_cache_show` could swallow it ;)   Benefit is not as much as shorter code as making it clear right away that result is a list which contains those items.

some obnoxious pythonista who came from functional languages could use 
```python
results = map(
  lambda out: parse_apt_cache_show_pkgs_output(utils.to_unicode(outs[0], "utf-8")),
  exec_gen
)
```
but imho it is ulgyish due to need of using lambda here ... if only we decoded outputs right away within execut_command_batch, then it could be just 
```python
results = map(parse_apt_cache_show_pkgs_output, exec_gen)
```
nice and clear.  Note: map in python2 returns a list and in python3 a generator.  but you could use `from six.moves import map` to guarantee it being a generator ;-)

Reply:
> you could express yourself via a list comprehension

If I'm reading it right, I think it'd take a little more work than
your example.  `results +=` is extending with a list rather than
appending an element, so there'd need to be flattening somewhere so
that your list comp didn't give a list of lists.

Reply:
The itertools docs also give `chain.from_iterable(listOfLists)` as a
snippet of flattening one level of nesting.

Reply:
So I had two cases: a sequence of dicts, and a sequence of lists.  To flatten the sequence of dicts I used: 

results = {k: v for d in results for k, v in d.items()}